cypress/integration/dashboard/dashboard_filtering_spec.js

```diff
@@ -0,0 +1,315 @@
+ * 3) Set the existing filter as pinned, re-check the visualizations
+ * 4) Remove the filter, and check the visualizations for proper updates
+ * 5) Create a new dashboard, and populate with a pie graph
+ * 6) Perform apply different filters on the pie graph and check the visualization for proper updates
+ * 7) Remove all filters and ensure the graph reverts to its original design
+ * 8) Test adding another pie graph to the dashboard and applying a filter to both grpahs
+ */
+
+describe('dashboard filtering', () => {
+  before(() => {
+    // Increase the viewport size to prevent pop-up notifications from blocking UI element
+    // E.g. after adding a saved visualization, a notification appears in the bottom right
+    // which could block clicking on a button located behind it.
+    cy.viewport(1900, 1080)
+
+    // TO DO: Setup the indices
+    // cy.request('PUT', 'localhost:9200/.kibana_1', cy.fixture('dashboard/mappings.json'))
+    // cy.request('POST', 'localhost:9200/.kibana_1', cy.fixture('dashboard/data.json.gz'))
+  })
+
+  after(() => {
+    // TO DO: Tear-down the indices
+    // cy.request('DELETE', 'localhost:9200')
+  })
+
+  describe('adding a filter that excludes all data', () => {
+    before(() => {
+      // Go to the Dashboards list page
+      cy.visit('app/dashboards/list')
+
+      // Click the "Create dashboard" button
+      cy.get('[data-test-subj="newItemButton"]', { timeout: 20000 }).should('be.visible').click()
+      cy.get('[data-test-subj="emptyDashboardWidget"]').should('be.visible')
+      // Change the time to be between Jan 1 2018 and Apr 13, 2018
+      cy.setDashboardDataRange('Apr 13, 2018 @ 00:00:00.000', 'Jan 1, 2018 @ 00:00:00.000')
+
+      // Add all "Filter Bytes Test" visualizations
+      cy.addDashboardPanels('Filter Bytes Test', 'visualization', true)
+
+      // Click the "Add" button and add all "Saved Searches"
+      cy.addDashboardPanels('Filter Bytes Test', 'search', false)
+
+      // Add filter
+      cy.addDashboardFilter('bytes', 'is', '12345678')
+
+      cy.get('[data-test-subj="filter filter-enabled filter-key-bytes filter-value-12,345,678 filter-unpinned "]').should('be.visible')
+    })
+    it('filters on pie charts', () => {
+      // Check that none of the pie charts are occupied with data (show "No results found")
+      cy.checkElementDoesNotExist('svg > g > g.arcs > path.slice')
+    })
+
+    it('area, bar and heatmap charts filtered', () => {
+      // Check that none of the charts are filled with data
+      cy.checkElementDoesNotExist('svg > g > g.series')
+    })
+
+    it('data tables are filtered', () => {
+      // Check that none of the data tables are filled with data
+      cy.checkElementDoesNotExist('[data-test-subj="paginated-table-body"] [data-cell-content]')
+    })
+
+    it('goal and guages are filtered', () => {
+      // Goal label should be 0, gauge label should be 0%
+      cy.checkElementContainsValue('svg > g > g > text.chart-label', 2, /^0%?$/)
```

Comment:
what does this `/^0%?$/` means, could we make it more readable

Reply:
`/^0%?$/` is a regular expression specifying a search pattern that looks for a string that only consists of "0", with an optional "%". The `cy.checkElementContainsValue` function checks whether the elements returned by the given selector contain a string that matches the regex pattern; in the case of this test, the function is checking that the goal visualization label displays "0" and that the gauge label display "0%" (utilizing that optional "%" character).

I've partially improved the readability by moving the creation of the regex into the function definition and having the function argument just be string of interest (e.g. `'0%?'` instead of `/^0%?$/`). However this is still not a perfect solution since the provided string must still contain any regex syntax (such as the "?" character to represent the optional flag), and a more comprehensive solution may involve avoiding regex all together.